{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Kipoi usage tutorial: Combined Predictions for a Patient\n",
    "\n",
    "This tutorial shows how to apply the model to a vcf file, if you would like to predict the *combined* effect of variants in a *single* patient.\n",
    "\n",
    "For this you need a (single-sample) vcf with the variants of this particular patient.\n",
    "\n",
    "The combined effect is interesting to check if a patient has several variants in the same 5'UTR that cooperate to lead to a strong variant effect. E.g. if your patient has two variants in the 5'UTR of a specific gene, which combine to form a new uTIS, this approach will detect that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import kipoi\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test run\n",
    "Make a test run to check that the environment is correctly configured and all required packages work correctly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OMP: Info #154: KMP_AFFINITY: Initial OS proc set respected: 27,31,59,63\n",
      "OMP: Info #214: KMP_AFFINITY: decoding x2APIC ids.\n",
      "OMP: Info #156: KMP_AFFINITY: 4 available OS procs\n",
      "OMP: Info #157: KMP_AFFINITY: Uniform topology\n",
      "OMP: Info #285: KMP_AFFINITY: topology layer \"LL cache\" is equivalent to \"socket\".\n",
      "OMP: Info #285: KMP_AFFINITY: topology layer \"L3 cache\" is equivalent to \"socket\".\n",
      "OMP: Info #285: KMP_AFFINITY: topology layer \"L2 cache\" is equivalent to \"core\".\n",
      "OMP: Info #285: KMP_AFFINITY: topology layer \"L1 cache\" is equivalent to \"core\".\n",
      "OMP: Info #191: KMP_AFFINITY: 1 socket x 2 cores/socket x 2 threads/core (2 total cores)\n",
      "OMP: Info #216: KMP_AFFINITY: OS proc to physical thread map:\n",
      "OMP: Info #171: KMP_AFFINITY: OS proc 27 maps to socket 3 core 6 thread 0 \n",
      "OMP: Info #171: KMP_AFFINITY: OS proc 59 maps to socket 3 core 6 thread 1 \n",
      "OMP: Info #171: KMP_AFFINITY: OS proc 31 maps to socket 3 core 7 thread 0 \n",
      "OMP: Info #171: KMP_AFFINITY: OS proc 63 maps to socket 3 core 7 thread 1 \n",
      "OMP: Info #252: KMP_AFFINITY: pid 484 tid 484 thread 0 bound to OS proc set 27\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.data]\u001b[0m successfully loaded the dataloader ../Framepool/. from /data/nasif12/home_if12/karollus/projects/models/Framepool/dataloader.py::StrandedSequenceVariantDataloader\u001b[0m\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.model]\u001b[0m Downloading model arguments weights from https://zenodo.org/record/3584238/files/Framepool_combined_residual.h5?download=1\u001b[0m\n",
      "Downloading https://zenodo.org/record/3584238/files/Framepool_combined_residual.h5?download=1 to /data/nasif12/home_if12/karollus/projects/models/Framepool/downloaded/model_files/weights/d1e9656725e730d509a09d5371e51bd2\n",
      "3.47MB [00:00, 4.14MB/s]                                                        \n",
      "Using TensorFlow backend.\n",
      "/opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_qint8 = np.dtype([(\"qint8\", np.int8, 1)])\n",
      "/opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:527: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_quint8 = np.dtype([(\"quint8\", np.uint8, 1)])\n",
      "/opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:528: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_qint16 = np.dtype([(\"qint16\", np.int16, 1)])\n",
      "/opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:529: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_quint16 = np.dtype([(\"quint16\", np.uint16, 1)])\n",
      "/opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:530: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_qint32 = np.dtype([(\"qint32\", np.int32, 1)])\n",
      "/opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:535: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  np_resource = np.dtype([(\"resource\", np.ubyte, 1)])\n",
      "2020-11-06 20:23:46.002534: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX\n",
      "2020-11-06 20:23:46.015474: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2200020000 Hz\n",
      "2020-11-06 20:23:46.015835: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x55cf3b676050 executing computations on platform Host. Devices:\n",
      "2020-11-06 20:23:46.015902: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>\n",
      "2020-11-06 20:23:46.016161: I tensorflow/core/common_runtime/process_util.cc:71] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.pipeline]\u001b[0m dataloader.output_schema is compatible with model.schema\u001b[0m\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.pipeline]\u001b[0m Initialized data generator. Running batches...\u001b[0m\n",
      "Downloading https://zenodo.org/record/3584238/files/gencodev19_5utr_sorted_noprefix.bed?download=1 to /data/nasif12/home_if12/karollus/projects/models/Framepool/downloaded/example_files/intervals_file\n",
      "3.87MB [00:00, 4.54MB/s]                                                        \n",
      "Downloading https://zenodo.org/record/3584238/files/Homo_sapiens.GRCh37.74.dna.chromosome.22.fa?download=1 to /data/nasif12/home_if12/karollus/projects/models/Framepool/downloaded/example_files/fasta_file\n",
      "52.2MB [00:02, 24.7MB/s]                                                        \n",
      "Downloading https://zenodo.org/record/3584238/files/chr22utrVar1000gen_sorted_nodup.vcf.gz?download=1 to /data/nasif12/home_if12/karollus/projects/models/Framepool/downloaded/example_files/vcf_file\n",
      "1.05MB [00:00, 1.52MB/s]                                                        \n",
      "Downloading https://zenodo.org/record/3584238/files/chr22utrVar1000gen_sorted_nodup.vcf.gz.tbi?download=1 to /data/nasif12/home_if12/karollus/projects/models/Framepool/downloaded/example_files/vcf_file_tbi\n",
      "8.19kB [00:00, 15.1kB/s]\n",
      "Downloading https://zenodo.org/record/3584238/files/chrom_order_noprefix.txt?download=1 to /data/nasif12/home_if12/karollus/projects/models/Framepool/downloaded/example_files/chr_order_file\n",
      "8.19kB [00:00, 15.7kB/s]\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.pipeline]\u001b[0m Returned data schema correct\u001b[0m\n",
      "  0%|                                                    | 0/36 [00:00<?, ?it/s]OMP: Info #252: KMP_AFFINITY: pid 484 tid 3447 thread 1 bound to OS proc set 31\n",
      "OMP: Info #252: KMP_AFFINITY: pid 484 tid 10357 thread 2 bound to OS proc set 59\n",
      "OMP: Info #252: KMP_AFFINITY: pid 484 tid 10358 thread 3 bound to OS proc set 63\n",
      "OMP: Info #252: KMP_AFFINITY: pid 484 tid 10359 thread 4 bound to OS proc set 27\n",
      "OMP: Info #252: KMP_AFFINITY: pid 484 tid 3446 thread 5 bound to OS proc set 31\n",
      "OMP: Info #252: KMP_AFFINITY: pid 484 tid 10654 thread 6 bound to OS proc set 59\n",
      "OMP: Info #252: KMP_AFFINITY: pid 484 tid 10655 thread 7 bound to OS proc set 63\n",
      "OMP: Info #252: KMP_AFFINITY: pid 484 tid 10656 thread 8 bound to OS proc set 27\n",
      "100%|███████████████████████████████████████████| 36/36 [01:41<00:00,  2.82s/it]\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.pipeline]\u001b[0m predict_example done!\u001b[0m\n",
      "\u001b[32mINFO\u001b[0m \u001b[44m[kipoi.cli.main]\u001b[0m Successfully ran test_predict\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!kipoi test \"../Framepool\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using downloaded and verified file: /data/nasif12/home_if12/karollus/projects/models/Framepool/downloaded/model_files/weights/d1e9656725e730d509a09d5371e51bd2\n",
      "WARNING:tensorflow:From /opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Colocations handled automatically by placer.\n",
      "WARNING:tensorflow:From /opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:3445: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.\n",
      "WARNING:tensorflow:From /opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/ops/math_ops.py:3066: to_int32 (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Use tf.cast instead.\n",
      "WARNING:tensorflow:From /opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/ops/math_grad.py:102: div (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Deprecated in favor of operator or tf.math.divide.\n"
     ]
    }
   ],
   "source": [
    "# Source model directly from directory\n",
    "model = kipoi.get_model(\"../Framepool\", source=\"dir\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optional: download example files and hg19 fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In case you want to run the example, you can download the example files using this code.\n",
    "If you have your own files, you can skip this step and simply adjust the paths in the next step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib.request\n",
    "import gzip\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make ExampleFile directory if it does not exist\n",
    "if not os.path.exists(\"ExampleFiles\"):\n",
    "    os.makedirs(\"ExampleFiles\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('ExampleFiles/hg19_idmap.tsv', <http.client.HTTPMessage at 0x2b6c9e5a0ac8>)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download vcf\n",
    "urllib.request.urlretrieve(\"https://zenodo.org/record/3584238/files/patho.vcf.gz?download=1\", 'ExampleFiles/patho.vcf.gz')\n",
    "# Download vcf tabix\n",
    "urllib.request.urlretrieve(\"https://zenodo.org/record/3584238/files/patho.vcf.gz.tbi?download=1\", 'ExampleFiles/patho.vcf.gz.tbi')\n",
    "# Download bed (with prefix)\n",
    "urllib.request.urlretrieve(\"https://zenodo.org/record/3584238/files/gencodev19_5utr_sorted.bed?download=1\", 'ExampleFiles/gencodev19_5utr_sorted.bed')\n",
    "# Download chromosome order file (with prefix)\n",
    "urllib.request.urlretrieve(\"https://zenodo.org/record/3584238/files/chrom_order.txt?download=1\", 'ExampleFiles/chrom_order.txt')\n",
    "# Download id mapping file\n",
    "urllib.request.urlretrieve(\"https://zenodo.org/record/3584238/files/hg19_idmap.tsv?download=1\", 'ExampleFiles/hg19_idmap.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download gzipped hg19 fasta (warning: 900mb)\n",
    "urllib.request.urlretrieve(\"https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/hg19.fa.gz\", 'ExampleFiles/hg19.fa.gz')\n",
    "# unzip\n",
    "with gzip.open('ExampleFiles/hg19.fa.gz', 'rb') as f_in:\n",
    "    with open('ExampleFiles/hg19.fa', 'wb') as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Provide the parameters\n",
    "This Dataloader requires the following input files:\n",
    "1. bed3+ where a specific user-specified column (>3, 1-based) of the bed denotes the strand\n",
    "   and a specific user-specified column (>3, 1-based) of the bed denotes the transcript id\n",
    "   (or some other id that explains which exons in the bed belong together to form one sequence).\n",
    "   All columns of the bed, except the first three, the id and the strand, are ignored. \n",
    "2. fasta file that provides the reference genome\n",
    "3. bgzip compressed (single sample) vcf that provides the variants\n",
    "4. A chromosome order file (such as a fai file) that specifies the order of chromosomes \n",
    "   (must be valid for all files)\n",
    "   \n",
    "The bed and vcf must both be sorted (by position) and a tabix index must be present (must lie in the same directory and have the same name + .tbi)\n",
    "\n",
    "The num_chr flag indicates whether chromosomes are listed numerically or with a chr prefix. This must be consistent across all input files!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**NB: Some common issues**\n",
    "1. Files do not agree with respect to the ordering or the chr prefix. This creates all kinds of problems.\n",
    "2. The vcf includes non-DNA characters, e.g. * or /. The model cannot currently account for this\n",
    "3. A multi-sample vcf is used, or one which mixes several individuals. This creates nonsensical results, since all variants that intersect a given 5'UTR are simultenously injected. Thus the predicted effects will not reflect the reality of one patient, but a mix of all of them. Split the vcf by patient and predict seperately for each\n",
    "4. There is more than one alternative allele for a given position in the vcf. In this case, currently, only the first is used, and the rest are ignored (as otherwise it is unclear what the variant effect is supposed to be). As a workaround, you can split into several vcf."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path of the vcf file\n",
    "vcf_path = \"ExampleFiles/patho.vcf.gz\"\n",
    "\n",
    "# Path of the fasta file\n",
    "fasta_path = \"ExampleFiles/hg19.fa\"\n",
    "# Set true if fasta has no chr prefix, false otherwise\n",
    "num_chr = False\n",
    "\n",
    "# Path of the bed file specifying human 5utr\n",
    "# For gencode 19, if chr prefix is present, use:\n",
    "# gencodev19_5utr_sorted.bed\n",
    "# Else: gencodev19_5utr_sorted_noprefix.bed (can be found in same zenodo repository)\n",
    "bed_path = \"ExampleFiles/gencodev19_5utr_sorted.bed\"\n",
    "id_column = 4\n",
    "\n",
    "# chromosome order file (the example file has the natural order)\n",
    "# a noprefix version can also be found at the Zenodo repository\n",
    "chr_order_path = \"ExampleFiles/chrom_order.txt\"\n",
    "\n",
    "# output file path\n",
    "output_file_path = \"patho.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run Prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [01:16<00:00, 76.46s/it]\n"
     ]
    }
   ],
   "source": [
    "model.pipeline.predict_to_file(output_file_path, {\"intervals_file\":bed_path, \n",
    "                               \"fasta_file\":fasta_path,\n",
    "                               \"vcf_file\":vcf_path,\n",
    "                               \"chr_order_file\": chr_order_path,\n",
    "                               \"id_column\":id_column,\n",
    "                               \"num_chr\":num_chr},\n",
    "                              batch_size=64)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data as dataframe\n",
    "df = pd.read_csv(output_file_path, sep=\"\\t\")\n",
    "df = df.rename(index=str, columns={\"metadata/chr\":\"chr\",\n",
    "          \"metadata/exon_positions\":\"exon_positions\",\n",
    "          \"metadata/id\":\"transcript_id\",\n",
    "          \"metadata/strand\":\"strand\",\n",
    "          \"metadata/variants\":\"variants\",\n",
    "          \"preds/mrl_fold_change\":\"mrl_fold_change\",\n",
    "          \"preds/shift_1\":\"shift_1\",\n",
    "          \"preds/shift_2\":\"shift_2\"}\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optional: id map to get a clearer output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# id map path\n",
    "# Provide an id mapping file to get a richer output:\n",
    "id_map_path = \"ExampleFiles/hg19_idmap.tsv\"\n",
    "\n",
    "# Id map\n",
    "df_map = pd.read_csv(id_map_path, sep=\"\\t\")\n",
    "df = df.merge(df_map, on=\"transcript_id\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Show results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each row of the results table corresponds to a transcript whose 5'UTR region (defined in the bed) intersects at least one variant in the vcf\n",
    "The table provides the following information:\n",
    "- exon_positions: the 5'UTR exons defining the 5'UTR of a specific transcript\n",
    "- transcript_id: the transcript id \n",
    "- strand\n",
    "- variants: the variants which were injected into this specific 5'UTR\n",
    "- mrl_fold_change: the predicted log2 fold change in mean ribosome load due to the injected variants\n",
    "- shift1: what the fold change would be if the canonical frame is shifted by one nt\n",
    "- shift2: what the fold cange would be if the canonical frame is shifted by two nt\n",
    "\n",
    "(if the mrl_fold_change has low magnitude, but one of the shifts has high magnitude, this indicates the creation/destruction of an in-frame uTIS that could lengthen/shorten the canonical protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>exon_positions</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>variants</th>\n",
       "      <th>mrl_fold_change</th>\n",
       "      <th>shift_1</th>\n",
       "      <th>shift_2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>19156944-19157295</td>\n",
       "      <td>ENST00000242261</td>\n",
       "      <td>-</td>\n",
       "      <td>chr7:19157207:G&gt;T;chr7:19157225:C&gt;A</td>\n",
       "      <td>-0.789764</td>\n",
       "      <td>-0.395907</td>\n",
       "      <td>-0.773174</td>\n",
       "      <td>ENSG00000122691</td>\n",
       "      <td>TWIST1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>96931119-96931250;96931606-96931732</td>\n",
       "      <td>ENST00000258439</td>\n",
       "      <td>-</td>\n",
       "      <td>chr2:96931137:G&gt;A</td>\n",
       "      <td>-1.285575</td>\n",
       "      <td>-1.012779</td>\n",
       "      <td>0.028387</td>\n",
       "      <td>ENSG00000135956</td>\n",
       "      <td>TMEM127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr3</td>\n",
       "      <td>98312348-98312567</td>\n",
       "      <td>ENST00000264193</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:98312358:C&gt;T</td>\n",
       "      <td>-1.006342</td>\n",
       "      <td>-0.834495</td>\n",
       "      <td>0.085337</td>\n",
       "      <td>ENSG00000080819</td>\n",
       "      <td>CPOX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr5</td>\n",
       "      <td>147211140-147211349</td>\n",
       "      <td>ENST00000296695</td>\n",
       "      <td>-</td>\n",
       "      <td>chr5:147211193:G&gt;A</td>\n",
       "      <td>-0.361474</td>\n",
       "      <td>-0.218929</td>\n",
       "      <td>-0.283831</td>\n",
       "      <td>ENSG00000164266</td>\n",
       "      <td>SPINK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21974826-21975097</td>\n",
       "      <td>ENST00000304494</td>\n",
       "      <td>-</td>\n",
       "      <td>chr9:21974860:C&gt;A</td>\n",
       "      <td>-0.910533</td>\n",
       "      <td>-0.525239</td>\n",
       "      <td>-1.060424</td>\n",
       "      <td>ENSG00000147889</td>\n",
       "      <td>CDKN2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr11</td>\n",
       "      <td>5248251-5248301</td>\n",
       "      <td>ENST00000335295</td>\n",
       "      <td>-</td>\n",
       "      <td>chr11:5248280:C&gt;T</td>\n",
       "      <td>-0.828528</td>\n",
       "      <td>0.006580</td>\n",
       "      <td>-0.919977</td>\n",
       "      <td>ENSG00000244734</td>\n",
       "      <td>HBB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>209974758-209974761;209975316-209975388;209979...</td>\n",
       "      <td>ENST00000367021</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:209975361:T&gt;A</td>\n",
       "      <td>-1.068000</td>\n",
       "      <td>-0.861996</td>\n",
       "      <td>0.055964</td>\n",
       "      <td>ENSG00000117595</td>\n",
       "      <td>IRF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>93297581-93297671</td>\n",
       "      <td>ENST00000370321</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:93297626:C&gt;A</td>\n",
       "      <td>-0.799757</td>\n",
       "      <td>-0.670084</td>\n",
       "      <td>0.025305</td>\n",
       "      <td>ENSG00000122406</td>\n",
       "      <td>RPL5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>66508542-66508720;66511534-66511540</td>\n",
       "      <td>ENST00000392711</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17:66508599:G&gt;A</td>\n",
       "      <td>-1.025860</td>\n",
       "      <td>0.015113</td>\n",
       "      <td>-1.124092</td>\n",
       "      <td>ENSG00000108946</td>\n",
       "      <td>PRKAR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr2</td>\n",
       "      <td>96931119-96931227;96931606-96931732</td>\n",
       "      <td>ENST00000432959</td>\n",
       "      <td>-</td>\n",
       "      <td>chr2:96931137:G&gt;A</td>\n",
       "      <td>-1.029097</td>\n",
       "      <td>-0.896816</td>\n",
       "      <td>-0.570132</td>\n",
       "      <td>ENSG00000135956</td>\n",
       "      <td>TMEM127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21974826-21975038</td>\n",
       "      <td>ENST00000446177</td>\n",
       "      <td>-</td>\n",
       "      <td>chr9:21974860:C&gt;A</td>\n",
       "      <td>-0.982123</td>\n",
       "      <td>-0.552211</td>\n",
       "      <td>-1.042604</td>\n",
       "      <td>ENSG00000147889</td>\n",
       "      <td>CDKN2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21974826-21974865</td>\n",
       "      <td>ENST00000498124</td>\n",
       "      <td>-</td>\n",
       "      <td>chr9:21974860:C&gt;A</td>\n",
       "      <td>-1.020021</td>\n",
       "      <td>-0.026780</td>\n",
       "      <td>-1.134101</td>\n",
       "      <td>ENSG00000147889</td>\n",
       "      <td>CDKN2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr5</td>\n",
       "      <td>147211140-147211198</td>\n",
       "      <td>ENST00000510027</td>\n",
       "      <td>-</td>\n",
       "      <td>chr5:147211193:G&gt;A</td>\n",
       "      <td>-0.818441</td>\n",
       "      <td>-0.669264</td>\n",
       "      <td>-0.730059</td>\n",
       "      <td>ENSG00000164266</td>\n",
       "      <td>SPINK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr17</td>\n",
       "      <td>66508590-66508689;66511127-66511260;66511534-6...</td>\n",
       "      <td>ENST00000536854</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17:66508599:G&gt;A</td>\n",
       "      <td>-0.142158</td>\n",
       "      <td>-0.277853</td>\n",
       "      <td>-0.063634</td>\n",
       "      <td>ENSG00000108946</td>\n",
       "      <td>PRKAR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr17</td>\n",
       "      <td>66508567-66508689;66511534-66511540</td>\n",
       "      <td>ENST00000589228</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17:66508599:G&gt;A</td>\n",
       "      <td>-1.129289</td>\n",
       "      <td>-0.975691</td>\n",
       "      <td>0.034666</td>\n",
       "      <td>ENSG00000108946</td>\n",
       "      <td>PRKAR1A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr                                     exon_positions    transcript_id  \\\n",
       "0    chr7                                  19156944-19157295  ENST00000242261   \n",
       "1    chr2                96931119-96931250;96931606-96931732  ENST00000258439   \n",
       "2    chr3                                  98312348-98312567  ENST00000264193   \n",
       "3    chr5                                147211140-147211349  ENST00000296695   \n",
       "4    chr9                                  21974826-21975097  ENST00000304494   \n",
       "5   chr11                                    5248251-5248301  ENST00000335295   \n",
       "6    chr1  209974758-209974761;209975316-209975388;209979...  ENST00000367021   \n",
       "7    chr1                                  93297581-93297671  ENST00000370321   \n",
       "8   chr17                66508542-66508720;66511534-66511540  ENST00000392711   \n",
       "9    chr2                96931119-96931227;96931606-96931732  ENST00000432959   \n",
       "10   chr9                                  21974826-21975038  ENST00000446177   \n",
       "11   chr9                                  21974826-21974865  ENST00000498124   \n",
       "12   chr5                                147211140-147211198  ENST00000510027   \n",
       "13  chr17  66508590-66508689;66511127-66511260;66511534-6...  ENST00000536854   \n",
       "14  chr17                66508567-66508689;66511534-66511540  ENST00000589228   \n",
       "\n",
       "   strand                             variants  mrl_fold_change   shift_1  \\\n",
       "0       -  chr7:19157207:G>T;chr7:19157225:C>A        -0.789764 -0.395907   \n",
       "1       -                    chr2:96931137:G>A        -1.285575 -1.012779   \n",
       "2       -                    chr3:98312358:C>T        -1.006342 -0.834495   \n",
       "3       -                   chr5:147211193:G>A        -0.361474 -0.218929   \n",
       "4       -                    chr9:21974860:C>A        -0.910533 -0.525239   \n",
       "5       -                    chr11:5248280:C>T        -0.828528  0.006580   \n",
       "6       -                   chr1:209975361:T>A        -1.068000 -0.861996   \n",
       "7       +                    chr1:93297626:C>A        -0.799757 -0.670084   \n",
       "8       +                   chr17:66508599:G>A        -1.025860  0.015113   \n",
       "9       -                    chr2:96931137:G>A        -1.029097 -0.896816   \n",
       "10      -                    chr9:21974860:C>A        -0.982123 -0.552211   \n",
       "11      -                    chr9:21974860:C>A        -1.020021 -0.026780   \n",
       "12      -                   chr5:147211193:G>A        -0.818441 -0.669264   \n",
       "13      +                   chr17:66508599:G>A        -0.142158 -0.277853   \n",
       "14      +                   chr17:66508599:G>A        -1.129289 -0.975691   \n",
       "\n",
       "     shift_2          gene_id gene_name  \n",
       "0  -0.773174  ENSG00000122691    TWIST1  \n",
       "1   0.028387  ENSG00000135956   TMEM127  \n",
       "2   0.085337  ENSG00000080819      CPOX  \n",
       "3  -0.283831  ENSG00000164266    SPINK1  \n",
       "4  -1.060424  ENSG00000147889    CDKN2A  \n",
       "5  -0.919977  ENSG00000244734       HBB  \n",
       "6   0.055964  ENSG00000117595      IRF6  \n",
       "7   0.025305  ENSG00000122406      RPL5  \n",
       "8  -1.124092  ENSG00000108946   PRKAR1A  \n",
       "9  -0.570132  ENSG00000135956   TMEM127  \n",
       "10 -1.042604  ENSG00000147889    CDKN2A  \n",
       "11 -1.134101  ENSG00000147889    CDKN2A  \n",
       "12 -0.730059  ENSG00000164266    SPINK1  \n",
       "13 -0.063634  ENSG00000108946   PRKAR1A  \n",
       "14  0.034666  ENSG00000108946   PRKAR1A  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter results\n",
    "1. Choose strongest result per gene (requires id mapping)\n",
    "2. Reduce to absolute effect > 0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>exon_positions</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>variants</th>\n",
       "      <th>mrl_fold_change</th>\n",
       "      <th>shift_1</th>\n",
       "      <th>shift_2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>abs_mrl_fc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>19156944-19157295</td>\n",
       "      <td>ENST00000242261</td>\n",
       "      <td>-</td>\n",
       "      <td>chr7:19157207:G&gt;T;chr7:19157225:C&gt;A</td>\n",
       "      <td>-0.789764</td>\n",
       "      <td>-0.395907</td>\n",
       "      <td>-0.773174</td>\n",
       "      <td>ENSG00000122691</td>\n",
       "      <td>TWIST1</td>\n",
       "      <td>0.789764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>96931119-96931250;96931606-96931732</td>\n",
       "      <td>ENST00000258439</td>\n",
       "      <td>-</td>\n",
       "      <td>chr2:96931137:G&gt;A</td>\n",
       "      <td>-1.285575</td>\n",
       "      <td>-1.012779</td>\n",
       "      <td>0.028387</td>\n",
       "      <td>ENSG00000135956</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>1.285575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr3</td>\n",
       "      <td>98312348-98312567</td>\n",
       "      <td>ENST00000264193</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:98312358:C&gt;T</td>\n",
       "      <td>-1.006342</td>\n",
       "      <td>-0.834495</td>\n",
       "      <td>0.085337</td>\n",
       "      <td>ENSG00000080819</td>\n",
       "      <td>CPOX</td>\n",
       "      <td>1.006342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr11</td>\n",
       "      <td>5248251-5248301</td>\n",
       "      <td>ENST00000335295</td>\n",
       "      <td>-</td>\n",
       "      <td>chr11:5248280:C&gt;T</td>\n",
       "      <td>-0.828528</td>\n",
       "      <td>0.006580</td>\n",
       "      <td>-0.919977</td>\n",
       "      <td>ENSG00000244734</td>\n",
       "      <td>HBB</td>\n",
       "      <td>0.828528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>209974758-209974761;209975316-209975388;209979...</td>\n",
       "      <td>ENST00000367021</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:209975361:T&gt;A</td>\n",
       "      <td>-1.068000</td>\n",
       "      <td>-0.861996</td>\n",
       "      <td>0.055964</td>\n",
       "      <td>ENSG00000117595</td>\n",
       "      <td>IRF6</td>\n",
       "      <td>1.068000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>93297581-93297671</td>\n",
       "      <td>ENST00000370321</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:93297626:C&gt;A</td>\n",
       "      <td>-0.799757</td>\n",
       "      <td>-0.670084</td>\n",
       "      <td>0.025305</td>\n",
       "      <td>ENSG00000122406</td>\n",
       "      <td>RPL5</td>\n",
       "      <td>0.799757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21974826-21974865</td>\n",
       "      <td>ENST00000498124</td>\n",
       "      <td>-</td>\n",
       "      <td>chr9:21974860:C&gt;A</td>\n",
       "      <td>-1.020021</td>\n",
       "      <td>-0.026780</td>\n",
       "      <td>-1.134101</td>\n",
       "      <td>ENSG00000147889</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>1.020021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr5</td>\n",
       "      <td>147211140-147211198</td>\n",
       "      <td>ENST00000510027</td>\n",
       "      <td>-</td>\n",
       "      <td>chr5:147211193:G&gt;A</td>\n",
       "      <td>-0.818441</td>\n",
       "      <td>-0.669264</td>\n",
       "      <td>-0.730059</td>\n",
       "      <td>ENSG00000164266</td>\n",
       "      <td>SPINK1</td>\n",
       "      <td>0.818441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr17</td>\n",
       "      <td>66508567-66508689;66511534-66511540</td>\n",
       "      <td>ENST00000589228</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17:66508599:G&gt;A</td>\n",
       "      <td>-1.129289</td>\n",
       "      <td>-0.975691</td>\n",
       "      <td>0.034666</td>\n",
       "      <td>ENSG00000108946</td>\n",
       "      <td>PRKAR1A</td>\n",
       "      <td>1.129289</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr                                     exon_positions    transcript_id  \\\n",
       "0    chr7                                  19156944-19157295  ENST00000242261   \n",
       "1    chr2                96931119-96931250;96931606-96931732  ENST00000258439   \n",
       "2    chr3                                  98312348-98312567  ENST00000264193   \n",
       "5   chr11                                    5248251-5248301  ENST00000335295   \n",
       "6    chr1  209974758-209974761;209975316-209975388;209979...  ENST00000367021   \n",
       "7    chr1                                  93297581-93297671  ENST00000370321   \n",
       "11   chr9                                  21974826-21974865  ENST00000498124   \n",
       "12   chr5                                147211140-147211198  ENST00000510027   \n",
       "14  chr17                66508567-66508689;66511534-66511540  ENST00000589228   \n",
       "\n",
       "   strand                             variants  mrl_fold_change   shift_1  \\\n",
       "0       -  chr7:19157207:G>T;chr7:19157225:C>A        -0.789764 -0.395907   \n",
       "1       -                    chr2:96931137:G>A        -1.285575 -1.012779   \n",
       "2       -                    chr3:98312358:C>T        -1.006342 -0.834495   \n",
       "5       -                    chr11:5248280:C>T        -0.828528  0.006580   \n",
       "6       -                   chr1:209975361:T>A        -1.068000 -0.861996   \n",
       "7       +                    chr1:93297626:C>A        -0.799757 -0.670084   \n",
       "11      -                    chr9:21974860:C>A        -1.020021 -0.026780   \n",
       "12      -                   chr5:147211193:G>A        -0.818441 -0.669264   \n",
       "14      +                   chr17:66508599:G>A        -1.129289 -0.975691   \n",
       "\n",
       "     shift_2          gene_id gene_name  abs_mrl_fc  \n",
       "0  -0.773174  ENSG00000122691    TWIST1    0.789764  \n",
       "1   0.028387  ENSG00000135956   TMEM127    1.285575  \n",
       "2   0.085337  ENSG00000080819      CPOX    1.006342  \n",
       "5  -0.919977  ENSG00000244734       HBB    0.828528  \n",
       "6   0.055964  ENSG00000117595      IRF6    1.068000  \n",
       "7   0.025305  ENSG00000122406      RPL5    0.799757  \n",
       "11 -1.134101  ENSG00000147889    CDKN2A    1.020021  \n",
       "12 -0.730059  ENSG00000164266    SPINK1    0.818441  \n",
       "14  0.034666  ENSG00000108946   PRKAR1A    1.129289  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"abs_mrl_fc\"] = np.abs(df[\"mrl_fold_change\"])\n",
    "idx = df.groupby(['gene_name'])['abs_mrl_fc'].transform(max) == df['abs_mrl_fc']\n",
    "df_max = df[idx]\n",
    "df_max[np.abs(df_max.mrl_fold_change) > 0.5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:anaconda-kipoi-Framepool2]",
   "language": "python",
   "name": "conda-env-anaconda-kipoi-Framepool2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
