{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Kipoi usage tutorial: Per-Variant Predictions\n",
    "\n",
    "This tutorial shows how to apply the model to a vcf file, if you would like to predict the *individual* effect of all variants in a vcf file.\n",
    "\n",
    "For this you need a vcf with all the variants of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import kipoi\n",
    "from kipoi.pipeline import Pipeline\n",
    "from kipoiseq.dataloaders import SingleVariantUTRDataLoader\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using downloaded and verified file: /data/nasif12/home_if12/karollus/projects/models/Framepool/downloaded/model_files/weights/d1e9656725e730d509a09d5371e51bd2\n",
      "WARNING:tensorflow:From /opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Colocations handled automatically by placer.\n",
      "WARNING:tensorflow:From /opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:3445: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.\n",
      "WARNING:tensorflow:From /opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/ops/math_ops.py:3066: to_int32 (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Use tf.cast instead.\n",
      "WARNING:tensorflow:From /opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/ops/math_grad.py:102: div (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Deprecated in favor of operator or tf.math.divide.\n"
     ]
    }
   ],
   "source": [
    "# Source model directly from directory\n",
    "model = kipoi.get_model(\"../Framepool\", source=\"dir\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optional: download example files and hg19 fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib.request\n",
    "import gzip\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make ExampleFile directory if it does not exist\n",
    "if not os.path.exists(\"ExampleFiles\"):\n",
    "    os.makedirs(\"ExampleFiles\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('ExampleFiles/hg19_idmap.tsv', <http.client.HTTPMessage at 0x2b3564049b70>)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download vcf\n",
    "urllib.request.urlretrieve(\"https://zenodo.org/record/3584238/files/patho.vcf.gz?download=1\", 'ExampleFiles/patho.vcf.gz')\n",
    "# Download vcf tabix\n",
    "urllib.request.urlretrieve(\"https://zenodo.org/record/3584238/files/patho.vcf.gz.tbi?download=1\", 'ExampleFiles/patho.vcf.gz.tbi')\n",
    "# Download GTF\n",
    "urllib.request.urlretrieve(\"https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/genes/hg19.ensGene.gtf.gz\", 'ExampleFiles/hg19.gtf.gz')\n",
    "# Download id mapping file\n",
    "urllib.request.urlretrieve(\"https://zenodo.org/record/3584238/files/hg19_idmap.tsv?download=1\", 'ExampleFiles/hg19_idmap.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download gzipped hg19 fasta (warning: 900mb)\n",
    "urllib.request.urlretrieve(\"https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/hg19.fa.gz\", 'ExampleFiles/hg19.fa.gz')\n",
    "# unzip\n",
    "with gzip.open('ExampleFiles/hg19.fa.gz', 'rb') as f_in:\n",
    "    with open('ExampleFiles/hg19.fa', 'wb') as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Provide the Parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This Dataloader requires the following input files:\n",
    "1. GTF which provides 5'UTR as a feature type (e.g. 5UTR in gencode, five_prime_utr in Ensembl)\n",
    "2. fasta file that provides the reference genome\n",
    "3. bgzip compressed (single sample) vcf that provides the variants\n",
    "   \n",
    "The vcf must be sorted (by position) and a tabix index must be present (must lie in the same directory and have the same name + .tbi)\n",
    "\n",
    "All files must agree on whether there is a chr prefix (e.g. is it chr1 or just 1)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**NB: Some common issues**\n",
    "1. Files do not agree with respect to the ordering or the chr prefix. This creates all kinds of problems.\n",
    "2. The vcf includes non-DNA characters, e.g. * or /. The model cannot currently account for this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path of the vcf file\n",
    "vcf_path = \"ExampleFiles/patho.vcf.gz\"\n",
    "\n",
    "# Path of the fasta file\n",
    "fasta_path = \"ExampleFiles/hg19.fa\"\n",
    "\n",
    "# Path of the gtf file\n",
    "gtf_path = \"ExampleFiles/hg19.gtf.gz\"\n",
    "# How are the 5'UTR called in your gtf\n",
    "feature_type = \"5UTR\"\n",
    "\n",
    "# output file path\n",
    "output_file_path = \"patho.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter your gtf (if necessary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If your vcf is small, or only covers small parts of the human genome, it is advisable to first subset your gtf to the corresponding regions to speed up the dataloader.\n",
    "\n",
    "Note: The dataloader was designed to handle very large vcf files (eg Gnomad VCF) that do not fit into memory easily. Thus, for very large VCF, filtering the gtf will not be beneficial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "from cyvcf2 import VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import gtf with pyranges\n",
    "gr = pr.read_gtf(gtf_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_set = set()\n",
    "for var in VCF(vcf_path):\n",
    "    chrom = var.CHROM\n",
    "    pos = var.POS\n",
    "    id_set = set(gr[chrom, pos-1:pos].df.gene_id) | id_set # vcf is 1-based, pyranges is not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "gr_subset = gr[gr.gene_id.isin(id_set)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_path = \"ExampleFiles/reduced.gtf\"\n",
    "gr_subset.to_gtf(gtf_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run Prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "pipeline = Pipeline(model, SingleVariantUTRDataLoader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1it [00:01,  1.10s/it]\n"
     ]
    }
   ],
   "source": [
    "pipeline.predict_to_file(output_file_path, {\"gtf_file\":gtf_path, \n",
    "                               \"fasta_file\":fasta_path,\n",
    "                               \"vcf_file\":vcf_path,\n",
    "                                \"feature_type\":feature_type},\n",
    "                              batch_size=64);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data as dataframe\n",
    "df = pd.read_csv(output_file_path, sep=\"\\t\")\n",
    "df = df.rename(index=str, columns={\"metadata/variant/chr\":\"chr\",\n",
    "          \"metadata/exon_positions\":\"exon_positions\",\n",
    "          \"metadata/transcript_id\":\"transcript_id\",\n",
    "          \"metadata/variant/str\":\"variants\",\n",
    "          \"metadata/variant/ref\":\"ref\",\n",
    "          \"metadata/variant/alt\":\"alt\",\n",
    "          \"metadata/variant/chrom\":\"chrom\",\n",
    "          \"metadata/variant/id\":\"id\",  \n",
    "          \"metadata/variant/pos\":\"pos\", \n",
    "          \"preds/mrl_fold_change\":\"mrl_fold_change\",\n",
    "          \"preds/shift_1\":\"shift_1\",\n",
    "          \"preds/shift_2\":\"shift_2\"}\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optional: id map to get a clearer output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# id map path\n",
    "# Provide an id mapping file to get a richer output:\n",
    "id_map_path = \"ExampleFiles/hg19_idmap.tsv\"\n",
    "\n",
    "# Id map\n",
    "df_map = pd.read_csv(id_map_path, sep=\"\\t\")\n",
    "df = df.merge(df_map, on=\"transcript_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>alt</th>\n",
       "      <th>chrom</th>\n",
       "      <th>id</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>variants</th>\n",
       "      <th>mrl_fold_change</th>\n",
       "      <th>shift_1</th>\n",
       "      <th>shift_2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000370321</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>rs376208311</td>\n",
       "      <td>93297626</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:93297626:C:A</td>\n",
       "      <td>-0.799757</td>\n",
       "      <td>-0.670084</td>\n",
       "      <td>0.025305</td>\n",
       "      <td>ENSG00000122406</td>\n",
       "      <td>RPL5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000470843</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>rs376208311</td>\n",
       "      <td>93297626</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:93297626:C:A</td>\n",
       "      <td>-0.107161</td>\n",
       "      <td>-0.189486</td>\n",
       "      <td>-0.054052</td>\n",
       "      <td>ENSG00000122406</td>\n",
       "      <td>RPL5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000367021</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CR022509</td>\n",
       "      <td>209975361</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1:209975361:T:A</td>\n",
       "      <td>-1.068000</td>\n",
       "      <td>-0.861996</td>\n",
       "      <td>0.055964</td>\n",
       "      <td>ENSG00000117595</td>\n",
       "      <td>IRF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000456314</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CR022509</td>\n",
       "      <td>209975361</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1:209975361:T:A</td>\n",
       "      <td>-0.675417</td>\n",
       "      <td>-0.531750</td>\n",
       "      <td>-0.228016</td>\n",
       "      <td>ENSG00000117595</td>\n",
       "      <td>IRF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000335295</td>\n",
       "      <td>T</td>\n",
       "      <td>chr11</td>\n",
       "      <td>rs34704828</td>\n",
       "      <td>5248280</td>\n",
       "      <td>C</td>\n",
       "      <td>chr11:5248280:C:T</td>\n",
       "      <td>-0.828528</td>\n",
       "      <td>0.006580</td>\n",
       "      <td>-0.919977</td>\n",
       "      <td>ENSG00000244734</td>\n",
       "      <td>HBB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000392711</td>\n",
       "      <td>A</td>\n",
       "      <td>chr17</td>\n",
       "      <td>CR023224</td>\n",
       "      <td>66508599</td>\n",
       "      <td>G</td>\n",
       "      <td>chr17:66508599:G:A</td>\n",
       "      <td>-1.025860</td>\n",
       "      <td>0.015113</td>\n",
       "      <td>-1.124092</td>\n",
       "      <td>ENSG00000108946</td>\n",
       "      <td>PRKAR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000585427</td>\n",
       "      <td>A</td>\n",
       "      <td>chr17</td>\n",
       "      <td>CR023224</td>\n",
       "      <td>66508599</td>\n",
       "      <td>G</td>\n",
       "      <td>chr17:66508599:G:A</td>\n",
       "      <td>-1.088306</td>\n",
       "      <td>-0.048745</td>\n",
       "      <td>-1.154857</td>\n",
       "      <td>ENSG00000108946</td>\n",
       "      <td>PRKAR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000585608</td>\n",
       "      <td>A</td>\n",
       "      <td>chr17</td>\n",
       "      <td>CR023224</td>\n",
       "      <td>66508599</td>\n",
       "      <td>G</td>\n",
       "      <td>chr17:66508599:G:A</td>\n",
       "      <td>-1.141302</td>\n",
       "      <td>-0.975095</td>\n",
       "      <td>0.025098</td>\n",
       "      <td>ENSG00000108946</td>\n",
       "      <td>PRKAR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000589228</td>\n",
       "      <td>A</td>\n",
       "      <td>chr17</td>\n",
       "      <td>CR023224</td>\n",
       "      <td>66508599</td>\n",
       "      <td>G</td>\n",
       "      <td>chr17:66508599:G:A</td>\n",
       "      <td>-1.129289</td>\n",
       "      <td>-0.975691</td>\n",
       "      <td>0.034666</td>\n",
       "      <td>ENSG00000108946</td>\n",
       "      <td>PRKAR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000536854</td>\n",
       "      <td>A</td>\n",
       "      <td>chr17</td>\n",
       "      <td>CR023224</td>\n",
       "      <td>66508599</td>\n",
       "      <td>G</td>\n",
       "      <td>chr17:66508599:G:A</td>\n",
       "      <td>-0.142158</td>\n",
       "      <td>-0.277853</td>\n",
       "      <td>-0.063634</td>\n",
       "      <td>ENSG00000108946</td>\n",
       "      <td>PRKAR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000392710</td>\n",
       "      <td>A</td>\n",
       "      <td>chr17</td>\n",
       "      <td>CR023224</td>\n",
       "      <td>66508599</td>\n",
       "      <td>G</td>\n",
       "      <td>chr17:66508599:G:A</td>\n",
       "      <td>-1.244144</td>\n",
       "      <td>-1.060489</td>\n",
       "      <td>-0.135118</td>\n",
       "      <td>ENSG00000108946</td>\n",
       "      <td>PRKAR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000258439</td>\n",
       "      <td>A</td>\n",
       "      <td>chr2</td>\n",
       "      <td>rs121908813</td>\n",
       "      <td>96931137</td>\n",
       "      <td>G</td>\n",
       "      <td>chr2:96931137:G:A</td>\n",
       "      <td>-1.285575</td>\n",
       "      <td>-1.012779</td>\n",
       "      <td>0.028387</td>\n",
       "      <td>ENSG00000135956</td>\n",
       "      <td>TMEM127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000432959</td>\n",
       "      <td>A</td>\n",
       "      <td>chr2</td>\n",
       "      <td>rs121908813</td>\n",
       "      <td>96931137</td>\n",
       "      <td>G</td>\n",
       "      <td>chr2:96931137:G:A</td>\n",
       "      <td>-1.029097</td>\n",
       "      <td>-0.896816</td>\n",
       "      <td>-0.570132</td>\n",
       "      <td>ENSG00000135956</td>\n",
       "      <td>TMEM127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000264193</td>\n",
       "      <td>T</td>\n",
       "      <td>chr3</td>\n",
       "      <td>rs867711777</td>\n",
       "      <td>98312358</td>\n",
       "      <td>C</td>\n",
       "      <td>chr3:98312358:C:T</td>\n",
       "      <td>-1.006342</td>\n",
       "      <td>-0.834495</td>\n",
       "      <td>0.085337</td>\n",
       "      <td>ENSG00000080819</td>\n",
       "      <td>CPOX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000513674</td>\n",
       "      <td>T</td>\n",
       "      <td>chr3</td>\n",
       "      <td>rs867711777</td>\n",
       "      <td>98312358</td>\n",
       "      <td>C</td>\n",
       "      <td>chr3:98312358:C:T</td>\n",
       "      <td>-0.989882</td>\n",
       "      <td>-0.828618</td>\n",
       "      <td>0.068531</td>\n",
       "      <td>ENSG00000080819</td>\n",
       "      <td>CPOX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000296695</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5</td>\n",
       "      <td>rs367798627</td>\n",
       "      <td>147211193</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:147211193:G:A</td>\n",
       "      <td>-0.361474</td>\n",
       "      <td>-0.218929</td>\n",
       "      <td>-0.283831</td>\n",
       "      <td>ENSG00000164266</td>\n",
       "      <td>SPINK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000510027</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5</td>\n",
       "      <td>rs367798627</td>\n",
       "      <td>147211193</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:147211193:G:A</td>\n",
       "      <td>-0.818441</td>\n",
       "      <td>-0.669264</td>\n",
       "      <td>-0.730059</td>\n",
       "      <td>ENSG00000164266</td>\n",
       "      <td>SPINK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000242261</td>\n",
       "      <td>T</td>\n",
       "      <td>chr7</td>\n",
       "      <td>30040876_1</td>\n",
       "      <td>19157207</td>\n",
       "      <td>G</td>\n",
       "      <td>chr7:19157207:G:T</td>\n",
       "      <td>-0.790121</td>\n",
       "      <td>-0.396197</td>\n",
       "      <td>-0.773752</td>\n",
       "      <td>ENSG00000122691</td>\n",
       "      <td>TWIST1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENST00000242261</td>\n",
       "      <td>A</td>\n",
       "      <td>chr7</td>\n",
       "      <td>30040876_2</td>\n",
       "      <td>19157225</td>\n",
       "      <td>C</td>\n",
       "      <td>chr7:19157225:C:A</td>\n",
       "      <td>0.000256</td>\n",
       "      <td>0.000234</td>\n",
       "      <td>0.000490</td>\n",
       "      <td>ENSG00000122691</td>\n",
       "      <td>TWIST1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENST00000304494</td>\n",
       "      <td>A</td>\n",
       "      <td>chr9</td>\n",
       "      <td>rs1800586</td>\n",
       "      <td>21974860</td>\n",
       "      <td>C</td>\n",
       "      <td>chr9:21974860:C:A</td>\n",
       "      <td>-0.910533</td>\n",
       "      <td>-0.525239</td>\n",
       "      <td>-1.060424</td>\n",
       "      <td>ENSG00000147889</td>\n",
       "      <td>CDKN2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENST00000498124</td>\n",
       "      <td>A</td>\n",
       "      <td>chr9</td>\n",
       "      <td>rs1800586</td>\n",
       "      <td>21974860</td>\n",
       "      <td>C</td>\n",
       "      <td>chr9:21974860:C:A</td>\n",
       "      <td>-1.020021</td>\n",
       "      <td>-0.026780</td>\n",
       "      <td>-1.134101</td>\n",
       "      <td>ENSG00000147889</td>\n",
       "      <td>CDKN2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENST00000446177</td>\n",
       "      <td>A</td>\n",
       "      <td>chr9</td>\n",
       "      <td>rs1800586</td>\n",
       "      <td>21974860</td>\n",
       "      <td>C</td>\n",
       "      <td>chr9:21974860:C:A</td>\n",
       "      <td>-0.982123</td>\n",
       "      <td>-0.552211</td>\n",
       "      <td>-1.042604</td>\n",
       "      <td>ENSG00000147889</td>\n",
       "      <td>CDKN2A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id alt  chrom           id        pos ref  \\\n",
       "0   ENST00000370321   A   chr1  rs376208311   93297626   C   \n",
       "1   ENST00000470843   A   chr1  rs376208311   93297626   C   \n",
       "2   ENST00000367021   A   chr1     CR022509  209975361   T   \n",
       "3   ENST00000456314   A   chr1     CR022509  209975361   T   \n",
       "4   ENST00000335295   T  chr11   rs34704828    5248280   C   \n",
       "5   ENST00000392711   A  chr17     CR023224   66508599   G   \n",
       "6   ENST00000585427   A  chr17     CR023224   66508599   G   \n",
       "7   ENST00000585608   A  chr17     CR023224   66508599   G   \n",
       "8   ENST00000589228   A  chr17     CR023224   66508599   G   \n",
       "9   ENST00000536854   A  chr17     CR023224   66508599   G   \n",
       "10  ENST00000392710   A  chr17     CR023224   66508599   G   \n",
       "11  ENST00000258439   A   chr2  rs121908813   96931137   G   \n",
       "12  ENST00000432959   A   chr2  rs121908813   96931137   G   \n",
       "13  ENST00000264193   T   chr3  rs867711777   98312358   C   \n",
       "14  ENST00000513674   T   chr3  rs867711777   98312358   C   \n",
       "15  ENST00000296695   A   chr5  rs367798627  147211193   G   \n",
       "16  ENST00000510027   A   chr5  rs367798627  147211193   G   \n",
       "17  ENST00000242261   T   chr7   30040876_1   19157207   G   \n",
       "18  ENST00000242261   A   chr7   30040876_2   19157225   C   \n",
       "19  ENST00000304494   A   chr9    rs1800586   21974860   C   \n",
       "20  ENST00000498124   A   chr9    rs1800586   21974860   C   \n",
       "21  ENST00000446177   A   chr9    rs1800586   21974860   C   \n",
       "\n",
       "              variants  mrl_fold_change   shift_1   shift_2          gene_id  \\\n",
       "0    chr1:93297626:C:A        -0.799757 -0.670084  0.025305  ENSG00000122406   \n",
       "1    chr1:93297626:C:A        -0.107161 -0.189486 -0.054052  ENSG00000122406   \n",
       "2   chr1:209975361:T:A        -1.068000 -0.861996  0.055964  ENSG00000117595   \n",
       "3   chr1:209975361:T:A        -0.675417 -0.531750 -0.228016  ENSG00000117595   \n",
       "4    chr11:5248280:C:T        -0.828528  0.006580 -0.919977  ENSG00000244734   \n",
       "5   chr17:66508599:G:A        -1.025860  0.015113 -1.124092  ENSG00000108946   \n",
       "6   chr17:66508599:G:A        -1.088306 -0.048745 -1.154857  ENSG00000108946   \n",
       "7   chr17:66508599:G:A        -1.141302 -0.975095  0.025098  ENSG00000108946   \n",
       "8   chr17:66508599:G:A        -1.129289 -0.975691  0.034666  ENSG00000108946   \n",
       "9   chr17:66508599:G:A        -0.142158 -0.277853 -0.063634  ENSG00000108946   \n",
       "10  chr17:66508599:G:A        -1.244144 -1.060489 -0.135118  ENSG00000108946   \n",
       "11   chr2:96931137:G:A        -1.285575 -1.012779  0.028387  ENSG00000135956   \n",
       "12   chr2:96931137:G:A        -1.029097 -0.896816 -0.570132  ENSG00000135956   \n",
       "13   chr3:98312358:C:T        -1.006342 -0.834495  0.085337  ENSG00000080819   \n",
       "14   chr3:98312358:C:T        -0.989882 -0.828618  0.068531  ENSG00000080819   \n",
       "15  chr5:147211193:G:A        -0.361474 -0.218929 -0.283831  ENSG00000164266   \n",
       "16  chr5:147211193:G:A        -0.818441 -0.669264 -0.730059  ENSG00000164266   \n",
       "17   chr7:19157207:G:T        -0.790121 -0.396197 -0.773752  ENSG00000122691   \n",
       "18   chr7:19157225:C:A         0.000256  0.000234  0.000490  ENSG00000122691   \n",
       "19   chr9:21974860:C:A        -0.910533 -0.525239 -1.060424  ENSG00000147889   \n",
       "20   chr9:21974860:C:A        -1.020021 -0.026780 -1.134101  ENSG00000147889   \n",
       "21   chr9:21974860:C:A        -0.982123 -0.552211 -1.042604  ENSG00000147889   \n",
       "\n",
       "   gene_name  \n",
       "0       RPL5  \n",
       "1       RPL5  \n",
       "2       IRF6  \n",
       "3       IRF6  \n",
       "4        HBB  \n",
       "5    PRKAR1A  \n",
       "6    PRKAR1A  \n",
       "7    PRKAR1A  \n",
       "8    PRKAR1A  \n",
       "9    PRKAR1A  \n",
       "10   PRKAR1A  \n",
       "11   TMEM127  \n",
       "12   TMEM127  \n",
       "13      CPOX  \n",
       "14      CPOX  \n",
       "15    SPINK1  \n",
       "16    SPINK1  \n",
       "17    TWIST1  \n",
       "18    TWIST1  \n",
       "19    CDKN2A  \n",
       "20    CDKN2A  \n",
       "21    CDKN2A  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter results\n",
    "1. Choose strongest result per gene (requires id mapping)\n",
    "2. Reduce to absolute effect > 0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>alt</th>\n",
       "      <th>chrom</th>\n",
       "      <th>id</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>variants</th>\n",
       "      <th>mrl_fold_change</th>\n",
       "      <th>shift_1</th>\n",
       "      <th>shift_2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>abs_mrl_fc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000370321</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>rs376208311</td>\n",
       "      <td>93297626</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:93297626:C:A</td>\n",
       "      <td>-0.799757</td>\n",
       "      <td>-0.670084</td>\n",
       "      <td>0.025305</td>\n",
       "      <td>ENSG00000122406</td>\n",
       "      <td>RPL5</td>\n",
       "      <td>0.799757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000367021</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>CR022509</td>\n",
       "      <td>209975361</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1:209975361:T:A</td>\n",
       "      <td>-1.068000</td>\n",
       "      <td>-0.861996</td>\n",
       "      <td>0.055964</td>\n",
       "      <td>ENSG00000117595</td>\n",
       "      <td>IRF6</td>\n",
       "      <td>1.068000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000335295</td>\n",
       "      <td>T</td>\n",
       "      <td>chr11</td>\n",
       "      <td>rs34704828</td>\n",
       "      <td>5248280</td>\n",
       "      <td>C</td>\n",
       "      <td>chr11:5248280:C:T</td>\n",
       "      <td>-0.828528</td>\n",
       "      <td>0.006580</td>\n",
       "      <td>-0.919977</td>\n",
       "      <td>ENSG00000244734</td>\n",
       "      <td>HBB</td>\n",
       "      <td>0.828528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000392710</td>\n",
       "      <td>A</td>\n",
       "      <td>chr17</td>\n",
       "      <td>CR023224</td>\n",
       "      <td>66508599</td>\n",
       "      <td>G</td>\n",
       "      <td>chr17:66508599:G:A</td>\n",
       "      <td>-1.244144</td>\n",
       "      <td>-1.060489</td>\n",
       "      <td>-0.135118</td>\n",
       "      <td>ENSG00000108946</td>\n",
       "      <td>PRKAR1A</td>\n",
       "      <td>1.244144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000258439</td>\n",
       "      <td>A</td>\n",
       "      <td>chr2</td>\n",
       "      <td>rs121908813</td>\n",
       "      <td>96931137</td>\n",
       "      <td>G</td>\n",
       "      <td>chr2:96931137:G:A</td>\n",
       "      <td>-1.285575</td>\n",
       "      <td>-1.012779</td>\n",
       "      <td>0.028387</td>\n",
       "      <td>ENSG00000135956</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>1.285575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000264193</td>\n",
       "      <td>T</td>\n",
       "      <td>chr3</td>\n",
       "      <td>rs867711777</td>\n",
       "      <td>98312358</td>\n",
       "      <td>C</td>\n",
       "      <td>chr3:98312358:C:T</td>\n",
       "      <td>-1.006342</td>\n",
       "      <td>-0.834495</td>\n",
       "      <td>0.085337</td>\n",
       "      <td>ENSG00000080819</td>\n",
       "      <td>CPOX</td>\n",
       "      <td>1.006342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000510027</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5</td>\n",
       "      <td>rs367798627</td>\n",
       "      <td>147211193</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:147211193:G:A</td>\n",
       "      <td>-0.818441</td>\n",
       "      <td>-0.669264</td>\n",
       "      <td>-0.730059</td>\n",
       "      <td>ENSG00000164266</td>\n",
       "      <td>SPINK1</td>\n",
       "      <td>0.818441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000242261</td>\n",
       "      <td>T</td>\n",
       "      <td>chr7</td>\n",
       "      <td>30040876_1</td>\n",
       "      <td>19157207</td>\n",
       "      <td>G</td>\n",
       "      <td>chr7:19157207:G:T</td>\n",
       "      <td>-0.790121</td>\n",
       "      <td>-0.396197</td>\n",
       "      <td>-0.773752</td>\n",
       "      <td>ENSG00000122691</td>\n",
       "      <td>TWIST1</td>\n",
       "      <td>0.790121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENST00000498124</td>\n",
       "      <td>A</td>\n",
       "      <td>chr9</td>\n",
       "      <td>rs1800586</td>\n",
       "      <td>21974860</td>\n",
       "      <td>C</td>\n",
       "      <td>chr9:21974860:C:A</td>\n",
       "      <td>-1.020021</td>\n",
       "      <td>-0.026780</td>\n",
       "      <td>-1.134101</td>\n",
       "      <td>ENSG00000147889</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>1.020021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_id alt  chrom           id        pos ref  \\\n",
       "0   ENST00000370321   A   chr1  rs376208311   93297626   C   \n",
       "2   ENST00000367021   A   chr1     CR022509  209975361   T   \n",
       "4   ENST00000335295   T  chr11   rs34704828    5248280   C   \n",
       "10  ENST00000392710   A  chr17     CR023224   66508599   G   \n",
       "11  ENST00000258439   A   chr2  rs121908813   96931137   G   \n",
       "13  ENST00000264193   T   chr3  rs867711777   98312358   C   \n",
       "16  ENST00000510027   A   chr5  rs367798627  147211193   G   \n",
       "17  ENST00000242261   T   chr7   30040876_1   19157207   G   \n",
       "20  ENST00000498124   A   chr9    rs1800586   21974860   C   \n",
       "\n",
       "              variants  mrl_fold_change   shift_1   shift_2          gene_id  \\\n",
       "0    chr1:93297626:C:A        -0.799757 -0.670084  0.025305  ENSG00000122406   \n",
       "2   chr1:209975361:T:A        -1.068000 -0.861996  0.055964  ENSG00000117595   \n",
       "4    chr11:5248280:C:T        -0.828528  0.006580 -0.919977  ENSG00000244734   \n",
       "10  chr17:66508599:G:A        -1.244144 -1.060489 -0.135118  ENSG00000108946   \n",
       "11   chr2:96931137:G:A        -1.285575 -1.012779  0.028387  ENSG00000135956   \n",
       "13   chr3:98312358:C:T        -1.006342 -0.834495  0.085337  ENSG00000080819   \n",
       "16  chr5:147211193:G:A        -0.818441 -0.669264 -0.730059  ENSG00000164266   \n",
       "17   chr7:19157207:G:T        -0.790121 -0.396197 -0.773752  ENSG00000122691   \n",
       "20   chr9:21974860:C:A        -1.020021 -0.026780 -1.134101  ENSG00000147889   \n",
       "\n",
       "   gene_name  abs_mrl_fc  \n",
       "0       RPL5    0.799757  \n",
       "2       IRF6    1.068000  \n",
       "4        HBB    0.828528  \n",
       "10   PRKAR1A    1.244144  \n",
       "11   TMEM127    1.285575  \n",
       "13      CPOX    1.006342  \n",
       "16    SPINK1    0.818441  \n",
       "17    TWIST1    0.790121  \n",
       "20    CDKN2A    1.020021  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"abs_mrl_fc\"] = np.abs(df[\"mrl_fold_change\"])\n",
    "idx = df.groupby(['gene_name'])['abs_mrl_fc'].transform(max) == df['abs_mrl_fc']\n",
    "df_max = df[idx]\n",
    "df_max[np.abs(df_max.mrl_fold_change) > 0.5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:anaconda-kipoi-Framepool2]",
   "language": "python",
   "name": "conda-env-anaconda-kipoi-Framepool2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
